[SYCL] Specialize atomic fetch_add for floating point types

AGindinson opened the pull request:
The new EXT/SPV_EXT_shader_atomic_float_add SPIR-V extension
allows us to further specialize `atomic::fetch_add()` for
floating point types. In device mode, we'll now be creating
an external call to a built-in-like `__spirv_AtomicFAddEXT()`.
This is similar to what is done for other atomic binary
instructions, e.g. the integer specialization of `fetch_add()`
being mapped onto `__spirv_AtomicIAdd()`.

Furthermore, `atomic::fetch_sub()` is also re-implemented
to use `__spirv_AtomicFAddEXT()`, the added operand being
a negation of the original one.

The new implementation can be exposed if a dedicated macro is
defined: `SYCL_USE_NATIVE_FP_ATOMICS`. Otherwise, a fallback
is used, where the atomic operation is done via spinlock emulation.
At the moment of committing this, only Intel GPUs support the
"native" implementation, which relies on a SPIR-V extension.

Tests for the feature have been finalized in
https://github.com/intel/llvm-test-suite/pull/104.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

AlexeySotkin commented on sycl/include/CL/sycl/ONEAPI/atomic_ref.hpp:
```diff
@@ -453,13 +453,17 @@ class atomic_ref_impl<
   T fetch_add(T operand, memory_order order = default_read_modify_write_order,
               memory_scope scope = default_scope) const noexcept {
+#if defined(__SYCL_DEVICE_ONLY__) && !defined(__NVPTX__)
+    return detail::spirv::AtomicFAdd(ptr, scope, order, operand);
```
Can we implement `fetch_sub()` as something like
```
auto Neg = detail::spirv::FNegate(operand); // not sure we have a thing like this.
return detail::spirv::AtomicFAdd(ptr, scope, order, Neg);
```
?

AlexeySachkov replied:
```
return detail::spirv::AtomicFAdd(ptr, scope, order, -operand);
```

AGindinson replied:
Done, at least as a draft.

Pennycook reviewed (approved):
The changes look good to me.

One quick question to make sure I understand: am I right in thinking that we now default to the native FAdd implementation for all our devices?  So the old compare-exchange implementation will now only be tested for CUDA devices?

@bader We should add a TODO somewhere to implement AtomicFAddEXT in the CUDA backend.

bader commented:
> We should add a TODO somewhere to implement AtomicFAddEXT in the CUDA backend.

I suggest adding a new issue with "cuda" tag - https://github.com/intel/llvm/issues.

AGindinson commented:
> One quick question to make sure I understand: am I right in thinking that we now default to the native FAdd implementation for all our devices? So the old compare-exchange implementation will now only be tested for CUDA devices?

Yes, this is aimed for all device BEs. I'll create an issue for CUDA once the [extension support](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/825) is pulled down into this repository.

AGindinson commented:
@Pennycook, could you please check if `__SYCL_EMULATE_NATIVE_ATOMICS__`  is a fine name for the macro (as introduced by https://github.com/intel/llvm/pull/2765/commits/d8712b500f0989abb651b4000f8e83f2c103d6ef)?

Also, should the IR checks in `test/atomic_ref/` be updated to include the emulated checks under the macro defined, or would it be sufficient to update github.com/intel/llvm-test-suite to use the macro for non-supporting targets and make sure all tests pass?

AGindinson commented:
/summary:run

AlexeySotkin commented on sycl/test/atomic_ref/sub.cpp:
```diff
@@ -165,22 +168,16 @@ void sub_test(queue q, size_t N) {
-  // CHECK-LLVM-SAME: (i32 addrspace(1)*, i32, i32, i32, i32, i32)
+  // CHECK-LLVM: declare dso_local spir_func float
+  // CHECK-LLVM-SAME: @_Z{{[0-9]+}}__spirv_AtomicFAddEXT
+  // CHECK-LLVM-SAME: (float addrspace(1)*, i32, i32, float)
```
Can we make sure that the argument is actually negated?

AGindinson replied:
In general, I would expect this to be checked within the actual test run in post-commit.
It basically comes down to how detailed we want the IR checks to be, just like in my earlier question:
> Also, should the IR checks in test/atomic_ref/ be updated to include the emulated checks under the macro defined?

The reason we want to have the existing checks for correct function calls is: "Make sure that not only GPU runs pass, but also that they use the native implementation, and not the emulated version that's known to work". A mere check for a correct declaration seems to fit this purpose. The reasoning for testing everything on the IR level (actual call instructions with correct arguments, including the negated ones; "emulated IR" layout under the emulation macro passed, etc.) would have more of a double-checking nature, and would also have to be modified once all BEs support the native atomics.

Given all that, what would you recommend as the best approach?

Pennycook commented:
> @Pennycook, could you please check if `__SYCL_EMULATE_NATIVE_ATOMICS__` is a fine name for the macro (as introduced by [d8712b5](https://github.com/intel/llvm/commit/d8712b500f0989abb651b4000f8e83f2c103d6ef))?

That commit says `__SYCL_EMULATE_FLOAT_ATOMICS__` rather than `__SYCL_EMULATE_NATIVE_ATOMICS__`.  But I think `__SYCL_EMULATE_FLOAT_ATOMICS__` is a good name.

> Also, should the IR checks in `test/atomic_ref/` be updated to include the emulated checks under the macro defined, or would it be sufficient to update github.com/intel/llvm-test-suite to use the macro for non-supporting targets and make sure all tests pass?

Good question.  If we're documenting `__SYCL_EMULATE_FLOAT_ATOMICS__` then I think we need to update the tests -- a device that implements `AtomicFAdd` would need to run correctly with and without the macro enabled, and so we should test that to ensure we don't accidentally break things.  If we're not documenting `__SYCL_EMULATE_FLOAT_ATOMICS__`, and it's only a temporary implementation detail, I think it's sufficient to ensure that the existing tests pass when the macro is set to the expected value.

AGindinson commented:
> That commit says __SYCL_EMULATE_FLOAT_ATOMICS__ rather than __SYCL_EMULATE_NATIVE_ATOMICS__. But I think __SYCL_EMULATE_FLOAT_ATOMICS__ is a good name.

That was an unfortunate typo on my side, sorry for that.
> A device that implements AtomicFAdd would need to run correctly with and without the macro enabled.

Fair enough. And yet I tend to agree that:
> it's only a temporary implementation detail

My personal approach would be to rely on the E2E tests passing with correct macro presence/absence. Should there be a lack of consensus among the reviewers, though, I'm prepared to lay this back and make all testing stages as detailed as possible. How would you recommend treating this? @AlexeySotkin, @AlexeySachkov, could you comment as well, please?

Pennycook commented on sycl/include/CL/sycl/atomic.hpp:
```diff
@@ -274,12 +280,20 @@ class atomic {
+        Ptr, SpirvScope, detail::getSPIRVMemorySemanticsMask(Order), Operand);
+  }
+
   T fetch_sub(T Operand, memory_order Order = memory_order::relaxed) {
```
Should you make the same change here as for `fetch_add`?  It might be a little surprising that we support `fetch_add` and `fetch_sub` in `atomic_ref`, but only `fetch_add` for `atomic`.

AGindinson replied:
Done; should be aligned properly now.

Pennycook commented on sycl/test/atomic_ref/add.cpp:
```diff
@@ -1,6 +1,7 @@
 // RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-device-only -S %s -o - \
 // RUN: | FileCheck %s --check-prefix=CHECK-LLVM
-// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out \
+// RUN: %gpu_atomics_config
```
The fact we have `RUN` after the line continuation looks odd to me.  I can see it's the same as the `|` on the lines above -- so it may be fine -- but I wanted to point this out so you can double-check it.

AGindinson replied:
Good note, I'll align this with https://github.com/intel/llvm-test-suite/pull/86.

AGindinson replied:
Done both here and in sub.cpp

Pennycook commented on sycl/test/atomic_ref/sub.cpp:
```diff
@@ -1,4 +1,7 @@
-// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-device-only -S %s -o - \
+// RUN: | FileCheck %s --check-prefix=CHECK-LLVM
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out \
+// RUN: %gpu_atomics_config
```
Same comment about line continuation as above.

Pennycook commented on sycl/include/CL/sycl/atomic.hpp:
```diff
@@ -274,18 +280,36 @@ class atomic {
+  template <typename T2 = T>
+  detail::enable_if_t<std::is_floating_point<T2>::value, T>
+  fetch_sub(T Operand, memory_order Order = memory_order::relaxed) {
+    // Negate the float add instruction.
+    return -__spirv_AtomicFAdd(
+        Ptr, SpirvScope, detail::getSPIRVMemorySemanticsMask(Order), Operand);
```
Should the negation here apply to `Operand` instead of the result?

AGindinson replied:
It **definitely** should, thanks for noting this :) Fixed.

That's what leaving a PR unattended for a month will do to me!

v-klochkov commented on sycl/test/lit.cfg.py:
```diff
@@ -85,6 +85,9 @@
 
 if triple == 'nvptx64-nvidia-cuda-sycldevice':
     config.available_features.add('cuda')
+    config.substitutions.append( ('%gpu_atomics_config', "-D__SYCL_EMULATE_FLOAT_ATOMICS__=1") )
```
Shouldn't this macro be set somewhere in clang driver?

AGindinson replied:
There are a couple points here:
1. The macro itself is more of a temporary fallback that can be used until all non-"Intel GPU" targets come to support the new implementation provided by the headers.
2. The targets that would want this configuration are not limited by NVPTX - for example, support for CPU is also lacking. However, the user would usually compile for the "SPIR target" in both CPU and Intel GPU cases, so we wouldn't be able to detect their intention to use emulated atomic functions at the Clang driver level.

That said, I really like your idea in a sense that CLI sugar like `-fsycl-emulate-float-atomics` could be provided by the Clang driver, and would improve UX even for the temporary purposes. This would require an independent PR, though. While I'll be able to prepare such a PR quickly, what would be your take on having such an improvement reviewed & merged after this main changeset?

AGindinson replied:
Suggested PR: https://github.com/intel/llvm/pull/3030

v-klochkov replied:
Thank you for the explanations, I though still have questions (do not consider them as stoppers for this PR).

What is 'nvptx64-nvidia-cuda-sycldevice' at the line 86 above? Isn't it some option or part of options to clang++ driver? If that is, why not to add that macro requiring emulation?
Ok, I guess that is doable, but then you say that CPU also does not support fast FP fetch_add, it only supports the emulation sequence. Wouldn't it be safer then to use the emulation sequence by default and true/fast FP fetch_add() by option like -fsycl-fast-float-atomics ?

AGindinson replied:
`nvptx64-*` is a valid target triple that is accepted by the Clang driver, exactly.

Basically, the primary target for the code containing such atomics are Intel GPUs, so these will be used in a majority of cases. Therefore, having this "new" behavior by default instead of the emulation sequence simplifies the usage for the main scenarios.

On the other hand, the goal with this new implementation of FP-typed `fetch_add`/`fetch_sub` is to have **all** of the targets eventually switched to accepting it. Therefore, adjustments for, say, NVPTX at the driver level would also be temporary - once https://github.com/intel/llvm/issues/2853 is resolved, we'd have to remove this internal driver setting. So a possible reason to avoid driver-level hard-coding is that it's longer to revert and release code changes than it is to stop using a particular macro/option on the client side. Again, with Intel GPUs as the primary target, this usage is likely to be modest.

AGindinson replied:
@v-klochkov, oh, now I'm finally getting your confusion as to this repo's lit configuration. The thing is, why do I even need to have an "emulated" compilation of the test if it's only launched on host?
I should simply remove this from here and only leave this configuration in the test-suite repo.

v-klochkov replied:
Just for the record: my questions to this place can be formulated as:
1) Are we Ok that users' programs that used FP fetch_add() and passed previously on CPU (with emulation sequence) will start failing after re-compilation with new headers?
2) dealing with such funcs as fetch_add becomes difficult without having good documentation telling how to check if FP fetch_add() is supported on particular device or not.

AGindinson replied:
As per an offline discussion, the current conclusion (which remains up for debate) appears to be:
1. We are OK with that, as we're moving towards all platforms support of the SPIR-V extensions employed for FP atomics.
2. The quick solution to "unfathomable runtime failures" after recompilation for unsupported targets (primarily, CPU) is that we add a Clang preprocessor warning to suggest a solution. The guide on supporting/non-supporting targets will be added to the documentation/release notes.

AGindinson commented:
Updates:
1. Switched to using the emulated implementation as default.
2. Renamed the macro correspondingly - it's `SYCL_USE_NATIVE_FP_ATOMICS` from now on. This is reflected in the test-suite PR: https://github.com/intel/llvm-test-suite/pull/104
3. Removed the changes from `atomic.hpp`, as this would've regressed the common `atomic` class for other targets. `atomic_ref` remains as the go-to class for native atomics.

romanovvlad reviewed (dismissed):
LGTM
